File: vhosts/blog/nuldoc-src/renderers/html.ts
import { Element, forEachChild, Node, Text } from "../dom.ts";
import { DocBookError } from "../errors.ts";

export function renderHtml(root: Node): string {
  return `<!DOCTYPE html>\n` + nodeToHtmlText(root, {
    indentLevel: -1,
    isInPre: false,
  });
}

type Context = {
  indentLevel: number;
  isInPre: boolean;
};

type Dtd = { type: "block" | "inline"; auto_closing?: boolean };

function getDtd(name: string): Dtd {
  switch (name) {
    case "__root__":
      return { type: "block" };
    case "a":
      return { type: "inline" };
    case "article":
      return { type: "block" };
    case "blockquote":
      return { type: "block" };
    case "body":
      return { type: "block" };
    case "br":
      return { type: "block", auto_closing: true };
    case "button":
      return { type: "block" };
    case "canvas":
      return { type: "block" };
    case "code":
      return { type: "inline" };
    case "div":
      return { type: "block" };
    case "em":
      return { type: "inline" };
    case "footer":
      return { type: "block" };
    case "h1":
      return { type: "inline" };
    case "h2":
      return { type: "inline" };
    case "h3":
      return { type: "inline" };
    case "h4":
      return { type: "inline" };
    case "h5":
      return { type: "inline" };
    case "h6":
      return { type: "inline" };
    case "head":
      return { type: "block" };
    case "header":
      return { type: "block" };
    case "hr":
      return { type: "block" };
    case "html":
      return { type: "block" };
    case "img":
      return { type: "block" };
    case "li":
      return { type: "block" };
    case "link":
      return { type: "block", auto_closing: true };
    case "main":
      return { type: "block" };
    case "meta":
      return { type: "block", auto_closing: true };
    case "nav":
      return { type: "block" };
    case "noscript":
      return { type: "block" };
    case "ol":
      return { type: "block" };
    case "p":
      return { type: "block" };
    case "pre":
      return { type: "block" };
    case "script":
      return { type: "block" };
    case "section":
      return { type: "block" };
    case "span":
      return { type: "inline" };
    case "strong":
      return { type: "inline" };
    case "sup":
      return { type: "inline" };
    case "table":
      return { type: "block" };
    case "tbody":
      return { type: "block" };
    case "td": // TODO
      return { type: "block" };
    case "tfoot":
      return { type: "block" };
    case "thead":
      return { type: "block" };
    case "time":
      return { type: "inline" };
    case "title": // TODO
      return { type: "inline" };
    case "tr":
      return { type: "block" };
    case "ul":
      return { type: "block" };
    default:
      throw new DocBookError(`[html.write] Unknown element name: ${name}`);
  }
}

function isInlineNode(n: Node): boolean {
  if (n.kind === "text") {
    return true;
  }
  if (n.name !== "a") {
    return getDtd(n.name).type === "inline";
  }

  // a tag: check if all children are inline elements.
  let allInline = true;
  forEachChild(n, (c) => allInline &&= isInlineNode(c));
  return allInline;
}

function isBlockNode(n: Node): boolean {
  return !isInlineNode(n);
}

function nodeToHtmlText(n: Node, ctx: Context): string {
  if (n.kind === "text") {
    if (n.raw) {
      return n.content;
    } else {
      return textNodeToHtmlText(n, ctx);
    }
  } else {
    return elementNodeToHtmlText(n, ctx);
  }
}

function textNodeToHtmlText(t: Text, ctx: Context): string {
  const s = encodeSpecialCharacters(t.content);
  if (ctx.isInPre) return s;

  // TODO: 日本語で改行するときはスペースを入れない
  return s.replaceAll(/\n */g, " ");
}

function encodeSpecialCharacters(s: string): string {
  return s.replaceAll(/&(?!\w+;)/g, "&amp;")
    .replaceAll(/</g, "&lt;")
    .replaceAll(/>/g, "&gt;")
    .replaceAll(/'/g, "&apos;")
    .replaceAll(/"/g, "&quot;");
}

function elementNodeToHtmlText(e: Element, ctx: Context): string {
  const dtd = getDtd(e.name);

  let s = "";
  if (e.name !== "__root__") {
    if (isBlockNode(e)) {
      s += indent(ctx);
    }
    s += `<${e.name}`;
    const attributes = getElementAttributes(e);
    if (attributes.length > 0) {
      s += " ";
      for (let i = 0; i < attributes.length; i++) {
        const [name, value] = attributes[i];
        s += `${name}="${value}"`;
        if (i !== attributes.length - 1) {
          s += " ";
        }
      }
    }
    s += ">";
    if (isBlockNode(e) && e.name !== "pre") {
      s += "\n";
    }
  }
  ctx.indentLevel += 1;

  let prevChild: Node | null = null;
  if (e.name === "pre") {
    ctx.isInPre = true;
  }
  forEachChild(e, (c) => {
    if (isBlockNode(e) && !ctx.isInPre) {
      if (isInlineNode(c)) {
        if (needsIndent(prevChild)) {
          s += indent(ctx);
        }
      } else {
        if (needsLineBreak(prevChild)) {
          s += "\n";
        }
      }
    }
    s += nodeToHtmlText(c, ctx);
    prevChild = c;
  });
  if (e.name === "pre") {
    ctx.isInPre = false;
  }

  ctx.indentLevel -= 1;
  if (e.name !== "__root__" && !dtd.auto_closing) {
    if (e.name !== "pre") {
      if (isBlockNode(e)) {
        if (needsLineBreak(prevChild)) {
          s += "\n";
        }
        s += indent(ctx);
      }
    }
    s += `</${e.name}>`;
    if (isBlockNode(e)) {
      s += "\n";
    }
  }
  return s;
}

function indent(ctx: Context): string {
  return "  ".repeat(ctx.indentLevel);
}

function getElementAttributes(e: Element): [string, string][] {
  return [...e.attributes.entries()]
    .filter((a) => !a[0].startsWith("--"))
    .sort(
      (a, b) => {
        // Special rules:
        if (e.name === "meta") {
          if (a[0] === "content" && b[0] === "name") {
            return 1;
          }
          if (a[0] === "name" && b[0] === "content") {
            return -1;
          }
        }
        if (e.name === "link") {
          if (a[0] === "href" && b[0] === "rel") {
            return 1;
          }
          if (a[0] === "rel" && b[0] === "href") {
            return -1;
          }
          if (a[0] === "href" && b[0] === "type") {
            return 1;
          }
          if (a[0] === "type" && b[0] === "href") {
            return -1;
          }
        }
        // General rules:
        if (a[0] > b[0]) return 1;
        else if (a[0] < b[0]) return -1;
        else return 0;
      },
    );
}

function needsIndent(prevChild: Node | null): boolean {
  return !prevChild || isBlockNode(prevChild);
}

function needsLineBreak(prevChild: Node | null): boolean {
  return !needsIndent(prevChild);
}

File: services/nuldoc/lib/nuldoc/renderers/html.rb
module Nuldoc
  class HtmlRenderer
    DTD = {
      'a' => { type: :inline },
      'article' => { type: :block },
      'blockquote' => { type: :block },
      'body' => { type: :block },
      'br' => { type: :block, self_closing: true },
      'button' => { type: :block },
      'canvas' => { type: :block },
      'caption' => { type: :block },
      'code' => { type: :inline },
      'del' => { type: :block },
      'div' => { type: :block },
      'em' => { type: :inline },
      'footer' => { type: :block },
      'h1' => { type: :inline },
      'h2' => { type: :inline },
      'h3' => { type: :inline },
      'h4' => { type: :inline },
      'h5' => { type: :inline },
      'h6' => { type: :inline },
      'head' => { type: :block },
      'header' => { type: :block },
      'hr' => { type: :block, self_closing: true },
      'html' => { type: :block },
      'i' => { type: :inline },
      'li' => { type: :block },
      'link' => { type: :block, self_closing: true },
      'img' => { type: :inline, self_closing: true },
      'ins' => { type: :inline },
      'main' => { type: :block },
      'mark' => { type: :inline },
      'meta' => { type: :block, self_closing: true },
      'nav' => { type: :block },
      'noscript' => { type: :block },
      'ol' => { type: :block },
      'p' => { type: :block },
      'pre' => { type: :block },
      'script' => { type: :block },
      'section' => { type: :block },
      'span' => { type: :inline },
      'strong' => { type: :inline },
      'sub' => { type: :inline },
      'sup' => { type: :inline },
      'table' => { type: :block },
      'tbody' => { type: :block },
      'td' => { type: :block },
      'tfoot' => { type: :block },
      'th' => { type: :block },
      'thead' => { type: :block },
      'time' => { type: :inline },
      'title' => { type: :inline },
      'tr' => { type: :block },
      'ul' => { type: :block },
      'svg' => { type: :block },
      'path' => { type: :block }
    }.freeze

    def render(root)
      "<!DOCTYPE html>\n#{node_to_html(root, indent_level: 0, is_in_pre: false)}"
    end

    private

    def get_dtd(name)
      dtd = DTD[name]
      raise "[html.write] Unknown element name: #{name}" if dtd.nil?

      dtd
    end

    def inline_node?(node)
      return true if %i[text raw].include?(node.kind)

      return get_dtd(node.name)[:type] == :inline if node.name != 'a'

      node.children.all? { |c| inline_node?(c) }
    end

    def block_node?(node)
      !inline_node?(node)
    end

    def node_to_html(node, indent_level:, is_in_pre:)
      case node.kind
      when :text
        text_node_to_html(node, _indent_level: indent_level, is_in_pre: is_in_pre)
      when :raw
        node.html
      when :element
        element_node_to_html(node, indent_level: indent_level, is_in_pre: is_in_pre)
      end
    end

    def text_node_to_html(node, _indent_level:, is_in_pre:)
      s = encode_special_characters(node.content)
      return s if is_in_pre

      s.gsub(/\n */) do |_match|
        offset = $LAST_MATCH_INFO.begin(0)
        last_char = offset.positive? ? s[offset - 1] : nil
        if ['。', '、'].include?(last_char)
          ''
        else
          ' '
        end
      end
    end

    def encode_special_characters(s)
      s.gsub(/&(?!(?:\w+|#\d+|#x[\da-fA-F]+);)/, '&amp;')
       .gsub('<', '&lt;')
       .gsub('>', '&gt;')
       .gsub("'", '&apos;')
       .gsub('"', '&quot;')
    end

    def element_node_to_html(element, indent_level:, is_in_pre:)
      dtd = get_dtd(element.name)
      s = +''

      s << indent(indent_level) if block_node?(element)
      s << "<#{element.name}"

      attributes = get_element_attributes(element)
      unless attributes.empty?
        s << ' '
        attributes.each_with_index do |(name, value), i|
          if name == 'defer' && value == 'true'
            s << 'defer'
          else
            attr_name = name == 'className' ? 'class' : name
            s << "#{attr_name}=\"#{encode_special_characters(value)}\""
          end
          s << ' ' if i != attributes.length - 1
        end
      end
      s << '>'
      s << "\n" if block_node?(element) && element.name != 'pre'

      child_indent = indent_level + 1
      prev_child = nil
      child_is_in_pre = is_in_pre || element.name == 'pre'

      element.children.each do |child|
        if block_node?(element) && !child_is_in_pre
          if inline_node?(child)
            s << indent(child_indent) if needs_indent?(prev_child)
          elsif needs_line_break?(prev_child)
            s << "\n"
          end
        end
        s << node_to_html(child, indent_level: child_indent, is_in_pre: child_is_in_pre)
        prev_child = child
      end

      unless dtd[:self_closing]
        if (element.name != 'pre') && block_node?(element)
          s << "\n" if needs_line_break?(prev_child)
          s << indent(indent_level)
        end
        s << "</#{element.name}>"
        s << "\n" if block_node?(element)
      end

      s
    end

    def indent(level)
      '  ' * level
    end

    def get_element_attributes(element)
      element.attributes
             .reject { |k, _| k.start_with?('__') }
             .compact
             .sort { |a, b| compare_attributes(element.name, a, b) }
    end

    def compare_attributes(element_name, a, b)
      ak, = a
      bk, = b

      if element_name == 'meta'
        return 1 if ak == 'content' && bk == 'name'
        return -1 if ak == 'name' && bk == 'content'
        return 1 if ak == 'content' && bk == 'property'
        return -1 if ak == 'property' && bk == 'content'
      end

      if element_name == 'link'
        return 1 if ak == 'href' && bk == 'rel'
        return -1 if ak == 'rel' && bk == 'href'
        return 1 if ak == 'href' && bk == 'type'
        return -1 if ak == 'type' && bk == 'href'
      end

      ak <=> bk
    end

    def needs_indent?(prev_child)
      prev_child.nil? || block_node?(prev_child)
    end

    def needs_line_break?(prev_child)
      !needs_indent?(prev_child)
    end
  end
end
